File: api/src/projectGlobalTasks/createPromoterTerminator.ts
/// <reference path="../@types/index.d.ts" />
import {Project, AnnotationPart, ProjectLog} from '../models'
const mongoose = require('mongoose');

const createPromoterTerminator = async ({_id, 
  promoterLength=500, 
  terminatorLength=200, 
  selectedRange,
}: {_id:any, promoterLength:number, terminatorLength:number, selectedRange?:IRange}, 
onProgress?:(progress:number, message:string)=>void,
) => {
  if (onProgress) {onProgress(5, 'task started')}
  const exists = await Project.exists({_id,});
  console.log(_id, exists);
  if(! exists) throw new Error(JSON.stringify({status:404, message: 'unable to find project'}));

const project = await Project
.findById(_id)
.exec();

let partCondition:any = {
  _id:{$in:project.parts},
  featureType: 'gene',
};
if (selectedRange) {
  partCondition = {...partCondition, start:{$gte: selectedRange.start}, end:{$lt: selectedRange.end}}
}

const partsLen = await AnnotationPart.find(partCondition)
.count();

const parts = await AnnotationPart.find(partCondition)
.sort({start:1, end:-1})
.cursor();

const replaceMap = {};

const projectLog:IProjectLog = {
_id:undefined,
conflictParts: [] as IPartUpdateLog[],
modifiedParts: [] as IPartUpdateLog[],
createdParts: [] as IPartUpdateLog[],
deletedParts: [] as IPartUpdateLog[],
shiftedParts: [] as IPartUpdateLog[],
}
let partsCount=0;
let lastPercentage = 0;

let timeStamp = Date.now();
await parts.eachAsync(async (part)=>{
  partsCount++;
  if(onProgress) {
    const percentage = 5 + Math.floor(90 * partsCount/partsLen);
    const now = Date.now();
    // only send every second and every percentage.
    if(percentage >= lastPercentage + 1 && now - timeStamp >= 1000) {
      onProgress(percentage, `now doing ${partsCount}/${partsLen}`);
      lastPercentage = percentage;
      timeStamp = now;
    }
  }
  let proStart:number;
  let proEnd:number;
  let terStart:number;
  let terEnd:number;
  const proName = part.name + '_P';
  const terName = part.name + '_T';
  const now = new Date();
  if (part.strand === -1) {
    proStart = part.end;
    proEnd = part.end + promoterLength;
    if (proEnd > project.len) proEnd = project.len;
    terEnd = part.start;
    terStart = part.start - terminatorLength;
    if (terStart < 0) terStart = 0;
  } else {
    proStart = part.start - promoterLength;
    proEnd = part.start;
    if (proStart < 0) proStart = 0;
    terStart = part.end;
    terEnd = part.end + terminatorLength;
    if (terEnd > project.len) terEnd = project.len;
  }
  const promoter = await AnnotationPart.create({
    pid: mongoose.Types.ObjectId(),
    featureType: 'promoter',
    chrId: part.chrId,
    chrName: part.chrName,
    start: proStart,
    end: proEnd,
    len: proEnd-proStart,
    strand: part.strand,
    name: proName,
    original: false,
    history: [],
    sequenceRef: {
      fileName: part.sequenceRef.fileName, 
      start: proStart, 
      end: proEnd, 
      strand: part.strand
    },
    built: false,
    parent: null,
    attribute: {ID:proName},
    createdAt:now,
    updatedAt:now,
    changelog: `created ${promoterLength}bp promoter for ${part.name}`,
  });

  projectLog.createdParts.push({
    ctype: 'new',
    part: promoter._id,
    name: promoter.name,
    changelog: promoter.changelog,
    location: promoter.start,
    oldPart: null,
  });

  const terminator = await AnnotationPart.create({
    pid: mongoose.Types.ObjectId(),
    featureType: 'terminator',
    chrId: part.chrId,
    chrName: part.chrName,
    start: terStart,
    end: terEnd,
    len: terEnd-terStart,
    strand: part.strand,
    name: terName,
    original: false,
    history: [],
    sequenceRef: {
      fileName: part.sequenceRef.fileName, 
      start: terStart, 
      end: terEnd, 
      strand: part.strand
    },
    built: false,
    parent: null,
    attribute: {ID:terName},
    createdAt:now,
    updatedAt:now,
    changelog: `created ${terminatorLength}bp terminator for ${part.name}`,
  });

  projectLog.createdParts.push({
    ctype: 'new',
    part: terminator._id,
    name: terminator.name,
    changelog: terminator.changelog,
    location: terminator.start,
    oldPart: null,
  });

  replaceMap[part._id.toString()] = part.strand === -1 ? [terminator._id, part._id, promoter._id] : [promoter._id, part._id, terminator._id];
});

const newPartIds = [];
// project.parts.forEach(v=>newPartIds.push(v));

project.parts.forEach(v=>{
  const replaceTarget = replaceMap[v.toString()];
  if(replaceTarget) {
    replaceTarget.forEach(vv=>newPartIds.push(vv));    
  } else {
    newPartIds.push(v);
  }
});

onProgress(95, `saving history`);

const projectObj = project.toObject();
const history = [{
  _id: projectObj._id,
  updatedAt: projectObj.updatedAt,
  changelog: projectObj.changelog,
}, ...projectObj.history];
delete projectObj._id;
const projectForm = {...projectObj, 
  parts: newPartIds,
  changelog: 'created promoter and terminators',
  history, updatedAt: new Date()};
const newProject = await Project.create(projectForm);
if (project.ctype !== 'source') {
  await Project.updateOne({_id:project._id}, {ctype:'history'});
}

projectLog._id = newProject._id;
await ProjectLog.create(projectLog);

return newProject;
}

export default createPromoterTerminator;

File: worker/algorithms/create_promoter_terminator.py
from pymongo import MongoClient
import pymongo
from bson.objectid import ObjectId
from algorithm_helpers.get_mongo_instance import get_mongo_instance
from algorithm_helpers.percentage import PercentageCounter
import datetime
import time
from munch import Munch, munchify


def create_promoter_terminator(project_id, promoter_length, terminator_length, **kwargs):
    db = get_mongo_instance()
    project_models = db.projects
    project = munchify(project_models.find_one({'_id':ObjectId(project_id)}))
    
    part_search_condition = {
        '_id':{'$in': project['parts']},
        'featureType': 'gene',
    }
    if 'start' in kwargs:
        part_search_condition['start'] = {'$gte': kwargs['start']}
    if 'end' in kwargs:
        part_search_condition['end'] =  {'$lt': kwargs['end']}    
    genes = db.annotation_parts.find(part_search_condition).sort([('start',pymongo.ASCENDING), ('end',pymongo.DESCENDING)])
    genes_len = genes.count()
    
    project_log = {
        '_id':None,
        'conflictParts': [],
        'modifiedParts': [],
        'createdParts': [],
        'deletedParts': [],
        'shiftedParts': [],
    }
    
    percentage_counter = PercentageCounter()
    percentage_counter.set_count_range(10, 90, genes_len)
    parts_count = 0
    
    now = datetime.datetime.now()
    new_part_list = []
    new_part_id_list = []
    
    replace_id_map = {}
    
    for part in genes:
        part = munchify(part)
        parts_count += 1
        progress_message = percentage_counter.build_message_by_count(parts_count, 'now doing {}/{}'.format(parts_count, genes_len))
        if progress_message:
            yield progress_message
        
        pro_name = part.name + '_P'
        ter_name = part.name + '_T'
        if part.strand == -1:
            pro_start = part.end
            pro_end = part.end + promoter_length
            if pro_end > project.len:
                pro_end = project.len
            ter_end = part.start
            ter_start = part.start - terminator_length
            if ter_start < 0:
                ter_start = 0
        else:
            pro_start = part.start - promoter_length
            pro_end = part.start
            if pro_start < 0:
                pro_start = 0
            ter_start = part.end
            ter_end = part.end + terminator_length
            if ter_end > project.len:
                ter_end = project.len
        # create parts
        promoter = Munch({
            '_id': ObjectId(),
            'pid': ObjectId(),
            'featureType': 'promoter',
            'chrId': part.chrId,
            'chrName': part.chrName,
            'start': pro_start,
            'end': pro_end,
            'len': pro_end - pro_start,
            'strand': part.strand,
            'name': pro_name,
            'original': False,
            'history': [],
            'sequenceRef': {
                'fileName': part.sequenceRef.fileName, 
                'start': pro_start, 
                'end': pro_end, 
                'strand': part.strand,
            },
            'built': False,
            'parent': None,
            'attribute': {'ID':pro_name},
            'createdAt':now,
            'updatedAt':now,
            'changelog': 'created {}bp promoter for {}'.format(promoter_length, part.name),
        })
        new_part_list.append(promoter.toDict())
        
        project_log['createdParts'].append({
            'ctype': 'new',
            'part': promoter._id,
            'name': promoter.name,
            'changelog': promoter.changelog,
            'location': promoter.start,
            'oldPart': None,
        })

        terminator = Munch({
            '_id': ObjectId(),
            'pid': ObjectId(),
            'featureType': 'terminator',
            'chrId': part.chrId,
            'chrName': part.chrName,
            'start': ter_start,
            'end': ter_end,
            'len': ter_end - ter_start,
            'strand': part.strand,
            'name': ter_name,
            'original': False,
            'history': [],
            'sequenceRef': {
                'fileName': part.sequenceRef.fileName, 
                'start': ter_start, 
                'end': ter_end, 
                'strand': part.strand,
            },
            'built': False,
            'parent': None,
            'attribute': {'ID':ter_name},
            'createdAt':now,
            'updatedAt':now,
            'changelog': 'created {}bp terminator for {}'.format(terminator_length, part.name),
        })
        new_part_list.append(terminator.toDict())
        
        project_log['createdParts'].append({
            'ctype': 'new',
            'part': terminator._id,
            'name': terminator.name,
            'changelog': terminator.changelog,
            'location': terminator.start,
            'oldPart': None,
        })
        if part.strand == -1:
            replace_id_map[part._id] = [terminator._id, part._id, promoter._id]
        else:
            replace_id_map[part._id] = [promoter._id, part._id, terminator._id]
        
        if len(new_part_list) >= 1000:
            x = db.abandon.insert_many(new_part_list)
            new_part_id_list += x.inserted_ids
            new_part_list = []
    if len(new_part_list) > 0:
        x = db.abandon.insert_many(new_part_list)
        new_part_id_list += x.inserted_ids
        new_part_list = []
        
    yield percentage_counter.build_message(92, 'saving history')
    
    new_parts = []
    for part_id in project.parts:
        if part_id in replace_id_map:
            new_parts += replace_id_map[part_id]
        else:
            new_parts.append(part_id)
    
    old_project_id = project._id
    old_project_ctype = project.ctype
    
    project.history = [{
        '_id': project._id,
        'updatedAt': project.updatedAt,
        'changelog': project.changelog,
    }] + project.history
    project._id = ObjectId()
    project.parts = new_parts
    project.ctype = 'project'
    project.changelog = 'created promoter and terminators'
    
    # create new project
    insert_result = db.projects.insert_one(project)
    if insert_result.acknowledged:
        # mark old project as history
        if old_project_ctype != 'source':
            modify_result = db.projects.update_one({'id':old_project_id}, { "$set": { "ctype": "history" } })
            if not modify_result.acknowledged:
                raise Exception('unable modify old project')
        yield percentage_counter.build_result(100, 'done', project._id)
    else:
        raise Exception('not inserted')
    
